test/common/ext_authz/ext_authz_impl_test.cc

```diff
@@ -0,0 +1,193 @@
+#include <chrono>
+#include <cstdint>
+#include <string>
+
+#include "common/ext_authz/ext_authz_impl.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+#include "common/network/address_impl.h"
+
+#include "test/mocks/grpc/mocks.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/network/mocks.h"
+#include "test/mocks/request_info/mocks.h"
+#include "test/mocks/ssl/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/test_common/printers.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::AtLeast;
+using testing::Invoke;
+using testing::Ref;
+using testing::Return;
+using testing::ReturnRef;
+using testing::WithArg;
+using testing::_;
+
+namespace Envoy {
+namespace ExtAuthz {
+
+class MockRequestCallbacks : public RequestCallbacks {
+public:
+  MOCK_METHOD1(onComplete, void(CheckStatus status));
+};
+
+class ExtAuthzGrpcClientTest : public testing::Test {
+public:
+  ExtAuthzGrpcClientTest()
+      : async_client_(new Grpc::MockAsyncClient()),
+        client_(Grpc::AsyncClientPtr{async_client_}, Optional<std::chrono::milliseconds>()) {}
+
+  Grpc::MockAsyncClient* async_client_;
+  Grpc::MockAsyncRequest async_request_;
+  GrpcClientImpl client_;
+  MockRequestCallbacks request_callbacks_;
+  Tracing::MockSpan span_;
+};
+
+TEST_F(ExtAuthzGrpcClientTest, Basic) {
+  std::unique_ptr<envoy::service::auth::v2::CheckResponse> response;
+
+  {
+    envoy::service::auth::v2::CheckRequest request;
+    Http::HeaderMapImpl headers;
+    EXPECT_CALL(*async_client_, send(_, ProtoEq(request), Ref(client_), _, _))
+        .WillOnce(Invoke([this](const Protobuf::MethodDescriptor& service_method,
+                                const Protobuf::Message&, Grpc::AsyncRequestCallbacks&,
+                                Tracing::Span&,
+                                const Optional<std::chrono::milliseconds>&) -> Grpc::AsyncRequest* {
+          EXPECT_EQ("envoy.service.auth.v2.Authorization", service_method.service()->full_name());
+          EXPECT_EQ("Check", service_method.name());
+          return &async_request_;
+        }));
+
+    client_.check(request_callbacks_, request, Tracing::NullSpan::instance());
+
+    client_.onCreateInitialMetadata(headers);
+    EXPECT_EQ(nullptr, headers.RequestId());
+
+    response.reset(new envoy::service::auth::v2::CheckResponse());
+    ::google::rpc::Status* status = new ::google::rpc::Status();
+    status->set_code(Grpc::Status::GrpcStatus::PermissionDenied);
+    response->set_allocated_status(status);
+    EXPECT_CALL(span_, setTag("ext_authz_status", "ext_authz_unauthorized"));
+    EXPECT_CALL(request_callbacks_, onComplete(CheckStatus::Denied));
+    client_.onSuccess(std::move(response), span_);
+  }
+
+  {
```

Comment:
Can you split this into multiple tests? The scoped-grouping-of-a-bunch-of-related-tests is fine for small simple tests, but I think here it would be clearer to have them as distinct tests.